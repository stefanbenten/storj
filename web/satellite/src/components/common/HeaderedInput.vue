// Copyright (C) 2018 Storj Labs, Inc.
// See LICENSE for copying information.

<template>
    <div class="input-container">
		<div v-if="!isOptional" class="label-container">
			<img v-if="error" src="../../../static/images/register/ErrorInfo.svg"/>
			<h3 v-if="!error">{{label}}</h3>
			<h3  v-if="!error" class="label-container__add-label">{{additionalLabel}}</h3>
			<h3 class="label-container__error" v-if="error">{{error}}</h3>
		</div>
		<div v-if="isOptional" class="optional-label-container">
			<h3>{{label}}</h3>
			<h4>Optional</h4>
		</div>
		<textarea
            v-if="isMultiline"
            :id="this.$props.label"
            :placeholder="this.$props.placeholder"
            :style="style"
            :rows="5"
            :cols="40"
            wrap="hard"
			v-model.lazy="value"
			@change="onInput"
            @input="onInput">
		</textarea>
		<input
            v-if="!isMultiline"
            :id="this.$props.label"
            :placeholder="this.$props.placeholder"
            v-bind:type="[isPassword ? 'password': 'text']"
            v-model.lazy="value"
            @change="onInput"
            @input="onInput"
            :style="style"/>
	</div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';

// Custom input component with labeled header
@Component(
    {
        data: function () {
            return {
                value: this.$props.initValue ? this.$props.initValue : '',
            };
        },
        methods: {
            // Emits data to parent component
            onInput() {
                this.$emit('setData', this.$data.value);
            },
            setValue(value: string) {
                this.$data.value = value;
            }
        },
        props: {
            initValue: {
                type: String,
            },
            label: {
                type: String,
                default: ''
            },
            additionalLabel: {
                type: String,
                default: ''
            },
            error: {
                type: String
            },
            placeholder: {
                type: String,
                default: 'default'
            },
            isOptional: {
                type: Boolean,
                default: false
            },
            isMultiline: {
                type: Boolean,
                default: false
            },
            isPassword: {
                type: Boolean,
                default: false
            },
            height: {
                type: String,
                default: '48px'
            },
            width: {
                type: String,
                default: '100%'
            },
        },
        computed: {
            style: function () {
                return {width: this.$props.width, height: this.$props.height};
            },
        },
    },
)

export default class HeaderedInput extends Vue {}

</script>

<style scoped lang="scss">
	.input-container {
		display: flex;
		flex-direction: column;
		align-items: flex-start;
		margin-top: 10px;
		width: 48%;
	}
	.label-container {
		display: flex;
		justify-content: flex-start;
		flex-direction: row;

		&__add-label {
			margin-left: 5px;
			font-family: 'montserrat_regular';
			font-size: 16px;
			line-height: 21px;
			color: rgba(56, 75, 101, 0.4);
		}

		&__error {
			color: #FF5560;
			margin-left: 10px;
		}
	}
	.optional-label-container {
		display: flex;
		flex-direction: row;
		justify-content: space-between;
		align-items: center;
		width: 100%;
		h4 {
			font-family: 'montserrat_regular';
			font-size: 16px;
			line-height: 21px;
			color: #AFB7C1;
		}
	}
	input,
	textarea {
		font-family: 'montserrat_regular';
		font-size: 16px;
		line-height: 21px;
		resize: none;
		height: 48px;
		width: 100%;
		text-indent: 20px;
		border-color: rgba(56, 75, 101, 0.4);
		border-radius: 6px;
		outline: none;
		box-shadow: none;
	}
	textarea {
		padding-top: 20px;
	}
	h3 {
		font-family: 'montserrat_regular';
		font-size: 16px;
		line-height: 21px;
		color: #354049;
	}
</style>
