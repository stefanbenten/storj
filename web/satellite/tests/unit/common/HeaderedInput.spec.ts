// Copyright (C) 2018 Storj Labs, Inc.
// See LICENSE for copying information.

import { mount, shallowMount } from '@vue/test-utils';
import HeaderedInput from '@/components/common/HeaderedInput.vue';

describe('HeaderedInput.vue', () => {

	it('renders correctly with default props', () => {

		const wrapper = shallowMount(HeaderedInput);

		expect(wrapper).toMatchSnapshot();
	});

	it('renders correctly with isMultiline props', () => {

		const wrapper = shallowMount(HeaderedInput, {
			propsData: {isMultiline: true}
		});

		expect(wrapper).toMatchSnapshot();
		expect(wrapper.contains('textarea')).toBe(true);
		expect(wrapper.contains('input')).toBe(false);
	});

	it('renders correctly with props', () => {
		let label = 'testLabel';
		let additionalLabel = 'addLabel';
		let width = '30px';
		let height = '20px';

		const wrapper = shallowMount(HeaderedInput, {
			propsData: {label, width, height, additionalLabel}
		});

		expect(wrapper.find('input').element.style.width).toMatch(width);
		expect(wrapper.find('input').element.style.height).toMatch(height);
		expect(wrapper.find('.label-container').text()).toMatch(label);
		expect(wrapper.find('.label-container__add-label').text()).toMatch(additionalLabel);
	});

	it('renders correctly with isOptional props', () => {

		const wrapper = shallowMount(HeaderedInput, {
			propsData: {
				isOptional: true
			}
		});

		expect(wrapper.find('h4').text()).toMatch('Optional');
	});

	it('renders correctly with input error', () => {
		let error = 'testError';

		const wrapper = shallowMount(HeaderedInput, {
			propsData: {
				error
			}
		});

		expect(wrapper).toMatchSnapshot();
		expect(wrapper.find('.label-container').text()).toMatch(error);
	});

	it('emit setData on input correctly', () => {
		let testData = 'testData';

		const wrapper = mount(HeaderedInput);

		wrapper.find('input').trigger('input');

		expect(wrapper.emitted('setData').length).toEqual(1);

		wrapper.vm.$emit('setData', testData);

		expect(wrapper.emitted('setData')[1][0]).toEqual(testData);
	});

});
