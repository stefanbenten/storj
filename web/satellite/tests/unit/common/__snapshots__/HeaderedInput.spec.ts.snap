// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderedInput.vue renders correctly with default props 1`] = `
<div class="input-container">
  <div class="label-container">
    <!---->
    <h3></h3>
    <h3 class="label-container__add-label"></h3>
    <!---->
  </div>
  <!---->
  <!----> <input id="" placeholder="default" type="text" style="width: 100%; height: 48px;">
</div>
`;

exports[`HeaderedInput.vue renders correctly with input error 1`] = `
<div class="input-container">
  <div class="label-container"><img src="../../../static/images/register/ErrorInfo.svg">
    <!---->
    <h3 class="label-container__add-label"></h3>
    <h3 class="label-container__error">testError</h3>
  </div>
  <!---->
  <!----> <input id="" placeholder="default" type="text" style="width: 100%; height: 48px;">
</div>
`;

exports[`HeaderedInput.vue renders correctly with isMultiline props 1`] = `
<div class="input-container">
  <div class="label-container">
    <!---->
    <h3></h3>
    <h3 class="label-container__add-label"></h3>
    <!---->
  </div>
  <!----> <textarea id="" placeholder="default" rows="5" cols="40" wrap="hard" style="width: 100%; height: 48px;"></textarea>
  <!---->
</div>
`;
